base58.1.scd
BASE58(1)

# NAME

base58 - Base58 encode or decode

# SYNOPSIS

*base58* [options...] [FILE]

# OPTIONS

*-h, --help*
	Show help message and exit.

*-d, --decode*
	Decode data instead of encoding

*-c, --check*
	Calculate a checksum and append to encoded data or verify existing checksum when decoding

# DESCRIPTION

Base58 encode or decode FILE (or standard input) to standard output. Optionally
a checksum can be used that is given by the first 4 bytes of the digest after
two passes of sha256.


CHANGELOG.md
## Changelog

### v2.1.1 / 2021-10-30
- [#67](https://github.com/keis/base58/pull/67) use github actions (@keis)
- [#66](https://github.com/keis/base58/pull/66) Escape illegal special characters in error message (@hukkin)

### v2.1.0 / 2021-01-09
- [#64](https://github.com/keis/base58/pull/64) Parametrise a few tests over alphabet (@keis)
- [#63](https://github.com/keis/base58/pull/63) Made it handle any base with passing alphabet. (#63) (@tanupoo)
- [#62](https://github.com/keis/base58/pull/62) Setup cfg (@keis)
- [#61](https://github.com/keis/base58/pull/61) Add support to ppc64le (@gururajrkatti)
- [#60](https://github.com/keis/base58/pull/60) Update README.md (@alloynetworks)
- [#59](https://github.com/keis/base58/pull/59) New alias for XRP alphabet (@alloynetworks)
- [#58](https://github.com/keis/base58/pull/58) Improve invalid character message (@keis)
- [#57](https://github.com/keis/base58/pull/57) Autofix for similar letters (@keis)
- [#56](https://github.com/keis/base58/pull/56) Add performance benchmarks using pytest-benchmark (@keis)
- [#53](https://github.com/keis/base58/pull/53) Performance optimizations (@kolomenkin)

### v2.0.1 / 2020-06-06
- [#55](https://github.com/keis/base58/pull/55) Include license file in source distribution (@synapticarbors)
- [#50](https://github.com/keis/base58/pull/50) Typecheck tests now that hamcrest has typing (@keis)
- [#49](https://github.com/keis/base58/pull/49) Replace custom exception assert with hamcrest utils (@keis)

### v2.0.0 / 2020-01-14
- [#47](https://github.com/keis/base58/pull/47) Distribute type data (PEP 561) (@hukkinj1)
- [#48](https://github.com/keis/base58/pull/48) Use setup.cfg for mypy conf (@hukkinj1)
- [#46](https://github.com/keis/base58/pull/46) Allow str input to b58encode_check (@keis)
- [#44](https://github.com/keis/base58/pull/44) Type annotate public API (@hukkinj1)
- [#43](https://github.com/keis/base58/pull/43) Drop python 2.7 and 3.4 support (#43) (@hukkinj1)
- [#42](https://github.com/keis/base58/pull/42) Test py3.7 and py3.8. Set python_requires in setup.py (@hukkinj1)
- [#41](https://github.com/keis/base58/pull/41) Fix encode functions' return types in README (@hukkinj1)
- [#40](https://github.com/keis/base58/pull/40) Enhance/add alphabet param (#40) (@dannywillems)
- [#39](https://github.com/keis/base58/pull/39) accept other bytes-like types (@fametrano)

### v1.0.3 / 2018-12-28
- [#37](https://github.com/keis/base58/pull/37) Update base58.py (@pdelteil)

### v1.0.2 / 2018-09-27
- [#34](https://github.com/keis/base58/pull/34) Add bumpversion config (@keis)

### v1.0.1 / 2018-09-25
- [#31](https://github.com/keis/base58/pull/31) Include tests in PyPI tarball (@dotlambda)

### v1.0.0 / 2018-04-20
- [#27](https://github.com/keis/base58/pull/27) Use bytes for both input and output (@keis)
- [#25](https://github.com/keis/base58/pull/25) Do not strip newline from input to cli (@keis)
- [#26](https://github.com/keis/base58/pull/26) Use tox and pytest to run tests (@keis)
- [#22](https://github.com/keis/base58/pull/22) Add the packaging metadata to build the base58 snap (@elopio)
- [#17](https://github.com/keis/base58/pull/17) improved error message (@fametrano)
- [#21](https://github.com/keis/base58/pull/21) clearer padding (@fametrano)

### v0.2.5 / 2017-04-04
- [#14](https://github.com/keis/base58/pull/14) Slight optimization, version increment (@gappleto97)
- [#12](https://github.com/keis/base58/pull/12) Add integer support (#12) (@gappleto97)
- [#10](https://github.com/keis/base58/pull/10) Add more test systems (@gappleto97)

### v0.2.4 / 2016-10-28
- [#8](https://github.com/keis/base58/pull/8) Package metadata (@keis)

### v0.2.3 / 2016-06-14
- [#6](https://github.com/keis/base58/pull/6) Improve error message when the type is not bytes (@keis, @zkanda)

### v0.2.2 / 2015-04-09
- [#3](https://github.com/keis/base58/pull/3) test round trips (@oconnor663)
- [#2](https://github.com/keis/base58/pull/2) fix encoding of empty string (@keis)


base58.py
'''Base58 encoding

Implementations of Base58 and Base58Check endcodings that are compatible
with the bitcoin network.
'''

# This module is based upon base58 snippets found scattered over many bitcoin
# tools written in python. From what I gather the original source is from a
# forum post by Gavin Andresen, so direct your praise to him.
# This module adds shiny packaging and support for python3.

from hashlib import sha256
from typing import Union

__version__ = '1.0.3'

# 58 character alphabet used
BITCOIN_ALPHABET = \
    b'123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
RIPPLE_ALPHABET = b'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'

# Retro compatibility
alphabet = BITCOIN_ALPHABET


def scrub_input(v: Union[str, bytes]) -> bytes:
    if isinstance(v, str):
        v = v.encode('ascii')

    return v


def b58encode_int(
    i: int, default_one: bool = True, alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode an integer using Base58
    """
    if not i and default_one:
        return alphabet[0:1]
    string = b""
    while i:
        i, idx = divmod(i, 58)
        string = alphabet[idx:idx+1] + string
    return string


def b58encode(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode a string using Base58
    """
    v = scrub_input(v)

    nPad = len(v)
    v = v.lstrip(b'\0')
    nPad -= len(v)

    p, acc = 1, 0
    for c in reversed(v):
        acc += p * c
        p = p << 8
    result = b58encode_int(acc, default_one=False, alphabet=alphabet)
    return alphabet[0:1] * nPad + result


def b58decode_int(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> int:
    """
    Decode a Base58 encoded string as an integer
    """
    v = v.rstrip()
    v = scrub_input(v)

    decimal = 0
    for char in v:
        decimal = decimal * 58 + alphabet.index(char)
    return decimal


def b58decode(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Decode a Base58 encoded string
    """
    v = v.rstrip()
    v = scrub_input(v)

    origlen = len(v)
    v = v.lstrip(alphabet[0:1])
    newlen = len(v)

    acc = b58decode_int(v, alphabet=alphabet)

    result = []
    while acc > 0:
        acc, mod = divmod(acc, 256)
        result.append(mod)

    return b'\0' * (origlen - newlen) + bytes(reversed(result))


def b58encode_check(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode a string using Base58 with a 4 character checksum
    """
    v = scrub_input(v)

    digest = sha256(sha256(v).digest()).digest()
    return b58encode(v + digest[:4], alphabet=alphabet)


def b58decode_check(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    '''Decode and verify the checksum of a Base58 encoded string'''

    result = b58decode(v, alphabet=alphabet)
    result, check = result[:-4], result[-4:]
    digest = sha256(sha256(result).digest()).digest()

    if check != digest[:4]:
        raise ValueError("Invalid checksum")

    return result


def main():
    '''Base58 encode or decode FILE, or standard input, to standard output.'''

    import sys
    import argparse

    stdout = sys.stdout.buffer

    parser = argparse.ArgumentParser(description=main.__doc__)
    parser.add_argument(
        'file',
        metavar='FILE',
        nargs='?',
        type=argparse.FileType('r'),
        default='-')
    parser.add_argument(
        '-d', '--decode',
        action='store_true',
        help='decode data')
    parser.add_argument(
        '-c', '--check',
        action='store_true',
        help='append a checksum before encoding')

    args = parser.parse_args()
    fun = {
        (False, False): b58encode,
        (False, True): b58encode_check,
        (True, False): b58decode,
        (True, True): b58decode_check
    }[(args.decode, args.check)]

    data = args.file.buffer.read()

    try:
        result = fun(data)
    except Exception as e:
        sys.exit(e)

    stdout.write(result)


if __name__ == '__main__':
    main()
